Doc/library/sqlite3.rst

```diff
@@ -1121,34 +1125,61 @@ Exceptions
 
       .. versionadded:: 3.11
 
+.. exception:: InterfaceError
+
+   Exception raised for misuse of the low-level SQLite C API.
+   In other words, if this exception is raised, it is probably a bug in the
+   ``sqlite3`` module.
+   ``InterfaceError`` is a subclass of :exc:`Error`.
+
 .. exception:: DatabaseError
 
    Exception raised for errors that are related to the database.
+   This serves the base exception for several types of database errors.
+   It is only raised implicitly through the specialised variants.
+   ``DatabaseError`` is a subclass of :exc:`Error`.
+
+.. exception:: DataError
+
+   Exception raised for errors caused by problems with the processed data,
+   like numeric value out of range, and too long strings.
+   ``DataError`` is a subclass of :exc:`DatabaseError`.
+
+.. exception:: OperationalError
+
+   Exception raised for errors that are related to the database's operation,
+   and not necessarily under the control of the programmer.
+   For example, the database path is not found,
+   or a transaction could not be processed.
+   ``OperationalError`` is a subclass of :exc:`DatabaseError`.
 
 .. exception:: IntegrityError
 
    Exception raised when the relational integrity of the database is affected,
-   e.g. a foreign key check fails.  It is a subclass of :exc:`DatabaseError`.
+   for example, a foreign key check fails.
+   ``IntegrityError`` is a subclass of :exc:`DatabaseError`.
 
-.. exception:: ProgrammingError
+.. exception:: InternalError
 
-   Exception raised for programming errors, e.g. table not found or already
-   exists, syntax error in the SQL statement, wrong number of parameters
-   specified, etc.  It is a subclass of :exc:`DatabaseError`.
+   Exception raised when SQLite encounters an internal error.
+   If this is raised, it may indicate that there is a problem with the runtime
+   SQLite library.
+   ``InternalError`` is a subclass of :exc:`DatabaseError`.
 
-.. exception:: OperationalError
+.. exception:: ProgrammingError
 
-   Exception raised for errors that are related to the database's operation
-   and not necessarily under the control of the programmer, e.g. an unexpected
-   disconnect occurs, the data source name is not found, a transaction could
-   not be processed, etc.  It is a subclass of :exc:`DatabaseError`.
+   Exception raised for ``sqlite3`` API programming errors.
+   for example syntax errors in SQL statements, and wrong number of bindings
+   supplied to a query.
+   ``ProgrammingError`` is a subclass of :exc:`DatabaseError`.
```

Comment:
Again, I feel like the original description of `ProgrammingError` was fine. This feels like it might be churn for not too much benefit.

Reply:
FTR, this is the original:

> Exception raised for programming errors, e.g. table not found or already
> exists, syntax error in the SQL statement, wrong number of parameters
> specified, etc.  It is a subclass of :exc:`DatabaseError`

I find "programming errors" vague; what are programming errors? SQL programming errors? Python sqlite3 API programming errors? sqlite3 SQLite C API programming errors? This is why I'm suggesting to replace it with something more conrecte: sqlite3 API programming errors. I'll see if I can phrase it better.

Reply:
Okay, that makes sense -- maybe just change that one line instead of reflowing the whole paragraph? Then it's more obvious to see what specific change you're making :)

Reply:
Hm, yeah, but looking at the original again, I see why I did this.

The first two examples are just wrong; both "table not found" and "table already exists" will result in `SQLITE_ERROR` which is mapped to `OperationalError`, not `ProgrammingError`.

The second example is also wrong, actually. Invalid input results in `OperationalError`, not `ProgrammingError`, so I have to remove that one.

The last example ("wrong number of parameters specified") is also slightly vague; parameters? SQL function parameters? Or Python function parameters? I think an accurate wording is better: "wrong number of _bindings_ supplied to a _query_".

Perhaps it is enought with one example?

Other examples of ProgrammingError are (this list is mostly a note to self):
- trying to use an uninitialised cursor or connection object
- trying to operate on a closed connection object
- returning unsupported values from a user-defined function (returning a type that SQLite does not natively support)
- trying use a connection or cursor object across different threads
- trying to set a connection limit for an unknown limit category
- recursive use of cursors
- binding parameters of unsupported types
- binding parameters by name, but no name was specified
- misssing bindings
- trying to execute non-DML statements using `executemany`

Reply:
I pushed a new wording for this.

Reply:
Thanks!